File: crates/vaultic-server/src/services/share_service.rs
//! Secure share service: create, retrieve (with view counting), delete.

use chrono::{Duration, Utc};
use rand::Rng;
use sea_orm::prelude::Expr;
use sea_orm::*;
use uuid::Uuid;

use crate::entities::secure_share;
use crate::error::AppError;

const ID_CHARS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
const SHARE_ID_LEN: usize = 12;

/// Generate a 12-char URL-safe random ID.
fn generate_share_id() -> String {
    let mut rng = rand::thread_rng();
    (0..SHARE_ID_LEN)
        .map(|_| ID_CHARS[rng.gen_range(0..ID_CHARS.len())] as char)
        .collect()
}

/// Create a new secure share link.
pub async fn create(
    db: &DatabaseConnection,
    user_id: Uuid,
    encrypted_data: &str,
    max_views: Option<i32>,
    ttl_hours: Option<i32>,
    vault_item_id: Option<Uuid>,
) -> Result<String, AppError> {
    let share_id = generate_share_id();
    let expires_at = ttl_hours.map(|h| (Utc::now() + Duration::hours(h as i64)).into());

    let share = secure_share::ActiveModel {
        id: Set(share_id.clone()),
        vault_item_id: Set(vault_item_id),
        user_id: Set(user_id),
        encrypted_data: Set(encrypted_data.to_string()),
        max_views: Set(max_views),
        current_views: Set(0),
        expires_at: Set(expires_at),
        created_at: Set(Some(Utc::now().into())),
        accessed_at: Set(None),
    };

    secure_share::Entity::insert(share).exec(db).await?;

    Ok(share_id)
}

/// Retrieve a share by ID using atomic view count increment to prevent TOCTOU races.
pub async fn retrieve(
    db: &DatabaseConnection,
    share_id: &str,
) -> Result<String, AppError> {
    let share = secure_share::Entity::find_by_id(share_id)
        .one(db)
        .await?
        .ok_or_else(|| AppError::NotFound("share not found".into()))?;

    // Check expiry
    if let Some(expires_at) = share.expires_at {
        let expires: chrono::DateTime<Utc> = expires_at.into();
        if Utc::now() > expires {
            return Err(AppError::Gone("share link has expired".into()));
        }
    }

    // Atomic increment: UPDATE ... SET current_views = current_views + 1
    // WHERE id = ? AND (max_views IS NULL OR current_views < max_views)
    let result = secure_share::Entity::update_many()
        .filter(secure_share::Column::Id.eq(share_id))
        .filter(
            Condition::any()
                .add(secure_share::Column::MaxViews.is_null())
                .add(
                    Expr::col(secure_share::Column::CurrentViews)
                        .lt(Expr::col(secure_share::Column::MaxViews)),
                ),
        )
        .col_expr(
            secure_share::Column::CurrentViews,
            Expr::col(secure_share::Column::CurrentViews).add(1),
        )
        .col_expr(
            secure_share::Column::AccessedAt,
            Expr::value(Utc::now()),
        )
        .exec(db)
        .await?;

    if result.rows_affected == 0 {
        return Err(AppError::Gone("share link max views reached".into()));
    }

    Ok(share.encrypted_data)
}

/// Get share metadata without counting a view.
pub async fn get_meta(
    db: &DatabaseConnection,
    share_id: &str,
) -> Result<(Option<i32>, i32, Option<chrono::DateTime<Utc>>), AppError> {
    let share = secure_share::Entity::find_by_id(share_id)
        .one(db)
        .await?
        .ok_or_else(|| AppError::NotFound("share not found".into()))?;

    // Check expiry
    if let Some(expires_at) = share.expires_at {
        let expires: chrono::DateTime<Utc> = expires_at.into();
        if Utc::now() > expires {
            return Err(AppError::Gone("share link has expired".into()));
        }
    }

    // Check max views reached
    if let Some(max) = share.max_views {
        if share.current_views >= max {
            return Err(AppError::Gone("share link max views reached".into()));
        }
    }

    let expires_at = share.expires_at.map(|e| e.into());
    Ok((share.max_views, share.current_views, expires_at))
}

/// Delete a share (owner only).
pub async fn delete(
    db: &DatabaseConnection,
    share_id: &str,
    user_id: Uuid,
) -> Result<(), AppError> {
    let share = secure_share::Entity::find_by_id(share_id)
        .one(db)
        .await?
        .ok_or_else(|| AppError::NotFound("share not found".into()))?;

    if share.user_id != user_id {
        return Err(AppError::Unauthorized("not the share owner".into()));
    }

    secure_share::Entity::delete_by_id(share_id).exec(db).await?;

    Ok(())
}

File: backend/src/services/share-service.ts
import { SecureShare } from "../models/secure-share-model.js";
import { AppError } from "../utils/app-error.js";

/** Create legacy share (with encrypted data on server). */
export async function create(
  userId: string,
  encryptedData: string,
  maxViews?: number,
  ttlHours?: number,
  vaultItemId?: string,
) {
  const expiresAt = ttlHours ? new Date(Date.now() + ttlHours * 3600_000) : null;

  const share = await SecureShare.create({
    userId,
    encryptedData,
    maxViews: maxViews ?? null,
    expiresAt,
    vaultItemId: vaultItemId ?? null,
  });

  return { share_id: share._id, expires_at: share.expiresAt };
}

/** Create metadata-only share (hybrid: data in URL, metadata on server). */
export async function createMetadata(
  userId: string | null,
  shareId: string,
  maxViews?: number,
  ttlHours?: number,
) {
  const expiresAt = ttlHours ? new Date(Date.now() + ttlHours * 3600_000) : null;

  // Check for duplicate share_id (client-provided) to prevent overwrite
  const existing = await SecureShare.findById(shareId).lean();
  if (existing) {
    throw AppError.conflict("share ID already exists — retry with a new link");
  }

  const share = await SecureShare.create({
    _id: shareId,
    userId: userId || "anonymous",
    encryptedData: null,
    maxViews: maxViews ?? null,
    expiresAt,
  });

  return { share_id: share._id, expires_at: share.expiresAt };
}

/**
 * Check share access — atomic view count + expiry check.
 * Returns { allowed: true } or throws with reason.
 */
export async function checkShareAccess(shareId: string) {
  const now = new Date();
  const share = await SecureShare.findOneAndUpdate(
    {
      _id: shareId,
      $and: [
        { $or: [{ maxViews: null }, { $expr: { $lt: ["$currentViews", "$maxViews"] } }] },
        { $or: [{ expiresAt: null }, { expiresAt: { $gt: now } }] },
      ],
    },
    { $inc: { currentViews: 1 }, $set: { accessedAt: now } },
    { new: true },
  );

  if (!share) {
    const existing = await SecureShare.findById(shareId).lean();
    if (!existing) throw AppError.notFound("share not found");
    if (existing.expiresAt && now > existing.expiresAt) {
      throw AppError.gone("share link has expired");
    }
    throw AppError.gone("share link max views reached");
  }

  return { allowed: true };
}

/**
 * Retrieve share data — atomic view counting via findOneAndUpdate.
 * For legacy shares that store encrypted data on server.
 */
export async function retrieve(shareId: string) {
  const now = new Date();
  const share = await SecureShare.findOneAndUpdate(
    {
      _id: shareId,
      $and: [
        { $or: [{ maxViews: null }, { $expr: { $lt: ["$currentViews", "$maxViews"] } }] },
        { $or: [{ expiresAt: null }, { expiresAt: { $gt: now } }] },
      ],
    },
    { $inc: { currentViews: 1 }, $set: { accessedAt: now } },
    { new: false },
  );

  if (!share) {
    const existing = await SecureShare.findById(shareId).lean();
    if (!existing) throw AppError.notFound("share not found");
    if (existing.expiresAt && now > existing.expiresAt) {
      throw AppError.gone("share link has expired");
    }
    throw AppError.gone("share link max views reached");
  }

  return { encrypted_data: share.encryptedData };
}

/**
 * Get share metadata — does NOT increment view count.
 */
export async function getMeta(shareId: string) {
  const share = await SecureShare.findById(shareId).lean();
  if (!share) throw AppError.notFound("share not found");

  const now = new Date();
  if (share.expiresAt && now > share.expiresAt) {
    throw AppError.gone("share link has expired");
  }
  if (share.maxViews !== null && share.currentViews >= share.maxViews) {
    throw AppError.gone("share link max views reached");
  }

  return {
    share_id: share._id,
    has_data: !!share.encryptedData, // true = legacy, false = hybrid
    max_views: share.maxViews,
    current_views: share.currentViews,
    expires_at: share.expiresAt,
    created_at: share.createdAt,
  };
}

/**
 * Delete share — owner only.
 */
export async function deleteShare(shareId: string, userId: string) {
  const share = await SecureShare.findById(shareId);
  if (!share) throw AppError.notFound("share not found");
  if (share.userId !== userId) throw AppError.unauthorized("not the share owner");

  await share.deleteOne();
  return { message: "share deleted" };
}
